### ISSUE TITLE
Batch Status Query Returns Non-Selective Query or 50,000 Row Limit Error in Customer Orgs with Thousands of Data Import Records

### ISSUE THREAD

**Issue by jesslopez-sf** [SalesforceFoundation/NPSP#5811]
The query throwing the error is in [BDI_DataImportBatchStatus_TDTM](https://github.com/SalesforceFoundation/NPSP/blob/61d3c4fd51ce07ab5f6a611f355ee6dbb5fb4cba/src/classes/BDI_DataImportBatchStatus_TDTM.cls), Lines 123-127.
```
SELECT Id, Batch_Status__c, 
	(SELECT Id, NPSP_Data_Import_Batch__c, Status__c FROM NPSP_Data_Imports__r) 
FROM DataImportBatch__c 
WHERE Id = :tmpVar1
```

**Issue 1:**
If you have more than 200,000 NPSP Data Import records in an org and try to Process a Batch:

```
npsp.TDTM_DataImport: execution of AfterInsert caused by: System.QueryException: Non-selective query against large object type (more than 200000 rows). Consider an indexed filter or contact salesforce.com about custom indexing. Even if a field is indexed a filter might still not be selective when: 1. The filter value includes null (for instance binding with a list that contains null) 2. Data skew exists whereby the number of matching rows is very large (for instance, filtering for a particular foreign key value that occurs many times) (npsp)
```
![Screen Shot 2020-09-04 at 1 31 12 PM](https://user-images.githubusercontent.com/29731456/92283743-904f2100-eece-11ea-8ff1-5e792197f9d5.png)

Other scenarios where I replicated:
* If you have more than 200,000 records associated with a batch where the Batch Status = ‘Completed’ and you attempt to add a new Data Import record. 
* If you have more than 200,000 Data Import records in general. And you have a Batch with Status of “Open” and > 50,000 Data Import records associated with it and you associate a new Data Import record with the batch  (Sometimes with fewer records as well…)
* I also replicated while Processing a Batch:
![Screen Shot 2020-09-04 at 1 37 30 PM](https://user-images.githubusercontent.com/29731456/92283772-9fce6a00-eece-11ea-986c-1801b217d6ef.png)

**Issue 2:** 
If you have more than 50,000 NPSP Data Import records associated with one NPSP Data Import Batch (but less than 200,000 Data Import records as a whole) and you try to insert another NPSP Data Import record tied to that batch, you hit:
`npsp.TDTM_DataImport: System.LimitException: npsp:Too many query rows: 50001`

----
Neither of these issues surfaced before adding the above query that we use to set the Batch Status field on the NPSP Data Import object. While it would be a best practice for customers to limit the total number of NPSP Data Import records associated to an NPSP Data Import Batch to 10,000 (per Salesforce documentation), customers may hit this now during their initial import (if data is tied to a batch) OR if they haven’t deleted Data Import records in the past.

### Workarounds:
1. Limit the number of NPSP Data Import records associated with an NPSP Data Import Batch to 10,000 or less.
2. Delete NPSP Data Import records that are no longer needed.

### Steps to Repeat:

**50,000 Row Error**
1. Insert 1 NPSP Data Import Batch record.
2. Insert 50,000 NPSP Data Import records associated with that Batch record. (You may need to disable the Trigger Handlers on the NPSP Data Import and NPSP Data Import Batch objects to avoid errors). Re-enable them when the 50,000 records are inserted.
3. Insert 1 more Data Import record associated with the NPSP Data Import Batch
4. Confirm the 50,000 row limit. 

**Non-selective query error:**
1. Disable the Trigger Handlers on the NPSP Data Import and NPSP Data Import Batch objects.
 2. Create a new NPSP Data Import Batch record.
3. Insert 200,000 NPSP Data Import records  and associate with that Batch record.
4. Re-enable the Trigger Handlers you disabled.
5. Go back to your original NPSP Data Import batch record (that has 50K records) and attempt to insert an NPSP Data Import record. You should receive the non-selective query error.

**Comment by jesslopez-sf** [comment]
W-8049403

**Comment by codexmarcel** [comment]
Included in production release 3.188
